Windows Overhaul: GTK 2.24, GStreaber OSS Build 10.7 beta, XULRunner 7.0

bendk opened the pull request:
I think this is finally ready to be really pulled in. It seems pretty stable and usable to me on my VM. Janet is going to test over the weakend, but I'd love some peer review as well.

Commit:
Initial gstreamer switch work

* updates setup.py to pull in gst libs from binary kit
* grabbed gst_extractor and gstreamerrenderer code from linux platform
* change file:// to file:/// urls
* removed VideoRenderer.shutdown() call from application

Commit:
Changes required for switching to pygtk all-in-one installer

* the gtk bundle is in a different place now, so this retools things
  to look for it there
* there are conflicting versions of a bunch of libraries between
  gtk and gstreamer, so this nixes grabbing the gstreamer ones

Commit:
Swaps out vlc extractor for gstreamer one

Commit:
Removes imagemagick bits

We don't use it anymore and haven't for a while.  Removing the remaining
bits from the installer.

Commit:
Adds audio track bits

Commit:
Fixes gtk DLL problems

GStreamer binaries that we use from OSSBuild come with gtk 2.16 dlls.  This
correctly sets the paths for things so that Miro picks up the right versions
of things and runs.

Commit:
Fixes gst plugin issues

* set environment variables so gstreamer can find the plugins
* put the plugins in a gstreamer-0.10 directory so they're separate
  from the other dlls--that way gstreamer doesn't try to load things
  it shouldn't when trying to find plugins

Commit:
Switches xid to handle

Commit:
Tweaks setup.py to handle development binary kits

This makes it a lot easier to make changes that involve binary kit
changes since it gets setup.py to look at the binary kit repository
if it's in the directory.

Commit:
Fixes problems launching downloader

The GST_* environment variables have unicode values.  We don't need to
pass them to the downloader, so I nixed them like FFMPEG_DATADIR.

Commit:
Rearranges startup, catches WindowsError

* rearranges startup so it's similar to how we do it on Linux with
  the figuring that these two platforms should be as close as possible
* we have exit_miro there because otherwise Miro doesn't shut down.
  however, it throws a WindowsError about an access violation.  given
  that we just want Miro to shut down, it seems ok to ignore the
  access violation.  it'd be good to figure out the problem at some
  point, though.

Commit:
Allow run.sh to work when "." is not in the path.

Commit:
Fixed threading issues on windows

We removed the threads_enter() and threads_leave() calls, which causes the GTK
thread not to run on my machine.  Instead of adding them back, switch to using
gobject.threads_init().  According to the PyGTK FAQ, this is what we want to
use when we have multiple threads, but only one of them acesses GTK.

Commit:
Refactored pygtkhacks a bit.

Changed get_gtk_widget() to get_c_gobject().  This makes it more general,
since it can work with any GObject.  We usually cast the result anyway so
there's no real difference for GtkWidgets.

Commit:
Make gstreamer work better on windows.

This changeset uses gdk_window_ensure_native() to give us a native HWND to
pass to gstreamer.  Before we were passing a client-side window to gstreamer
which confused it and made it take up the whole window area.

Events don't seem to be working correctly.  Both double-click and motion
events aren't firing on windows.

I also added a pygtkhacks function to calculate where inside a native window a
client-side window is, but it's not used yet.  I think it will be useful for
making XULRunner work again, and we may even end up using it for gstreamer as
well.

Commit:
Moved gstreamer code to portable

Made a miro.frontends.widgets.gst package that contains portable gstreamer
code.  Split up the renderer and sniffer code.

Made a portable make_file_url() method, that works on both linux and windows.
This was almost the only difference between the gstreamer code on both
platforms.

The gst package still has some GTK-specific bits to it.  I plan to remove
these soon.

Commit:
Refactored how we embed gstreamer.

Put the code that handles embedding gstreamer inside 1 platform-specific
class.  This is in preparation of overhauling the windows version.

I also tried to generally clean up the code.  We now clean up VideoWidget
properly, so we don't have to keep around a global singleton for it.
Simplified the Renderer interface.  Now all the renderer care's about is a
window id to pass it it's imagesink.  It doesn't need to connect to the
expose/destroy signals.

Commit:
Fixed window that embeds gstreamer on windows.

Stopped using GDKWindows and started just making a native window ourselves.
Added a C extension module that handles the Win32 functions.  Made the
extension module somewhat re-usable.  I think we might start using it for
XULRunner as well.

This fixes motion events, double-click events, and seems more stable to me.

Commit:
add DLL directories to PATH so py2exe finds them

Commit:
don't include extra DLLs

Commit:
pyflakes

Commit:
Use GPL version of OSSBuild gstreamer 10.7 beta 4

This works in conjunction with a miro-binary-kit-win change to fix 2 issues
with audio on windows

The main issue was that AAC audio was not supported.  Using the GPL builds
adds the faad element, which decodes AAC.

The other issue was that the dsoundsink would only work with 44100 khz audio.
I'm not sure what the problem was, but changing to the GPL version seems to
have fixed it.

Change the code to pick our audio/video sinks to be platform-specific.  On
windows, we know which sinks we want to use, so we don't need to have all the
testing code that linux uses.  On linux, we just use the old code from
portable.

Changed how we construct AudioRenderer and VideoRenderer.  Instead of having
the renderer module define the clases, it defines a factory method that
creates the objects we need.  This is more flexible and it was the simplest
way to have different sink picking code on different platforms.

Commit:
Refactored gst renderer.

We may as use reset() for that code, if the function is definined.

Commit:
Switch to XULRunner 7.0 on windows

Updated our C++ code to work with XULRunner 7.0.  XULRunner isn't working yet
after the GTK changes, but the module compiles.

Commit:
Use EmbeddingWindow for the XULRunner browser

This makes XULRunner work again.

Now that both the browser and renderer code use embeddingwindow, created
embeddingwidget, wraps it inside a GTK widget and handles common tasks.

Added code to destroy leftover windows like we used to do with
PersistentWindow.  Added code to SiteDisplay to explicitly destroy browsers
when we are through with them.

Fixed WM_PAINT handling code for embedding window.  Calling DefWindowProc
makes it so that windows marks the region as clean.  Otherwise we just keep
getting WM_PAINT messages.

Removed FixFocus.cpp and replaced it with code in embeddingwindow.c.  Found
the SetWindowSubclass() function, which is much cleaner than manually handling
things with SetWindowLong.

Commit:
Fixed fullscreen bug after overhaul.

I originally changed this because it fixed a crash with the overhaul code.
However, after some other fixes, it's not needed anymore.  I still think it's
a bit cleaner and safer than the other code though.

Commit:
Fixed NSIS installer.

Updated it to reflect the overhaul changes.

paulswartz commented:
I can't review this at the moment (don't have a working Windows VM), but if it hasn't gotten a review when my computer gets back, I will.

paulswartz commented:
Working on this now.  First thing I notice is that `binary_kit_version` hasn't been updated (understandably).  That will need to get done before this gets merged.

Commit:
Fixed subtitle font selection on windows (#15227)

Now that we use gstreamer, font selection is way simpler and works correctly
with unicode.  We just use pango to get the list of font families to select.

Moved renderer code to select fonts to portable.  On liunx, there's no UI to
change SUBTITLE_FONT, but it doesn't hurt to have it there.  Users could use
gconf-editor to change it if they really wanted to.

Commit:
Updating the binary kit version in preparation of the merge.

paulswartz commented:
Build went off fine, so I'm merging this in.
